crowdsourcing/serializers/payment.py

```diff
@@ -0,0 +1,162 @@
+from crowdsourcing.validators.utils import InequalityValidator, ConditionallyRequiredValidator
+from crowdsourcing.utils import PayPalBackend, get_model_or_none
+from rest_framework import status
+from django.utils import timezone
+
+class FinancialAccountSerializer(DynamicFieldsModelSerializer):
+    class Meta:
+        model = FinancialAccount
+        fields = ('id', 'owner', 'type', 'is_active', 'balance')
+
+
+class TransactionSerializer(DynamicFieldsModelSerializer):
+    class Meta:
+        model = Transaction
+        fields = ('id', 'sender_type', 'amount', 'currency', 'state', 'method',
+                  'sender', 'recipient', 'reference', 'created_timestamp', 'last_updated')
+        read_only_fields = ('created_timestamp', 'last_updated')
+
+    def create(self, *args, **kwargs):
+        transaction = Transaction.objects.create(**self.validated_data)
+        transaction.recipient.balance += transaction.amount
+        transaction.recipient.save()
+        if not transaction.sender.is_system:
+            transaction.sender.balance -= transaction.amount
+            transaction.sender.save()
+        return transaction
+
+
+class PayPalFlowSerializer(DynamicFieldsModelSerializer):
+    class Meta:
+        model = PayPalFlow
+        fields = ('id', 'paypal_id', 'state', 'recipient', 'redirect_url', 'payer_id')
+        read_only_fields = ('state', 'recipient')
+
+    def create(self, *args, **kwargs):
+        flow = PayPalFlow()
+        flow.state = 'created'
+        flow.recipient = kwargs['recipient']
+        flow.paypal_id = self.validated_data['paypal_id']
+        flow.redirect_url = self.validated_data['redirect_url']
+        flow.save()
+        return flow
+
+    def execute(self, *args, **kwargs):
+        paypalbackend = PayPalBackend()
+        payment = paypalbackend.paypalrestsdk.Payment.find(self.validated_data['paypal_id'])
+        if payment.execute({"payer_id": self.validated_data['payer_id']}):
+            flow = PayPalFlow.objects.get(paypal_id=self.validated_data['paypal_id'])
+            flow.state='approved'
+            flow.payer_id=self.validated_data['payer_id']
+            flow.save()
+            transaction = {
+                "amount": payment["transactions"][0]["amount"]["total"],
+                "currency": payment["transactions"][0]["amount"]["currency"],
+                "recipient": flow.recipient.id,
+                "reference": payment["id"],
+                "state": "approved",
+                "method": payment["payer"]["payment_method"],
+                "sender": FinancialAccount.objects.get(is_system=True, type="paypal_deposit").id
+            }
+            if not self.context['request'].user.is_anonymous():
+                transaction["sender_type"] = "self"
+            else:
+                transaction["sender_type"] = "other"
+            serializer = TransactionSerializer(data=transaction)
+            if serializer.is_valid():
```

Comment:
If this fails for some reason, we need to make sure we end up crediting the requester's account later.